spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,133 @@
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> partitionIdx;
+  private final boolean markUpdatedRows;
+
+  private Row cachedRow = null;
+
+  public ChangelogIterator(
+      Iterator<Row> rowIterator,
+      int changeTypeIndex,
+      List<Integer> partitionIdx,
+      boolean markUpdatedRows) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.partitionIdx = partitionIdx;
+    this.markUpdatedRows = markUpdatedRows;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is a processed cached row, return it directly
+    if (cachedRow != null
+        && !cachedRow.getString(changeTypeIndex).equals(DELETE)
+        && !cachedRow.getString(changeTypeIndex).equals(INSERT)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+
+      if (withinPartition(currentRow, nextRow)
+          && currentRow.getString(changeTypeIndex).equals(DELETE)
+          && nextRow.getString(changeTypeIndex).equals(INSERT)) {
+
+        GenericInternalRow deletedRow =
+            new GenericInternalRow(((GenericRowWithSchema) currentRow).values());
+        GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+        // set the change_type to the same value
+        deletedRow.update(changeTypeIndex, "");
+        insertedRow.update(changeTypeIndex, "");
+
+        if (deletedRow.equals(insertedRow)) {
+          // remove two carry-over rows
+          currentRow = null;
+          this.cachedRow = null;
+        } else if (markUpdatedRows) {
+          // mark the updated rows
+          deletedRow.update(changeTypeIndex, ChangelogOperation.UPDATE_BEFORE.name());
+          currentRow = RowFactory.create(deletedRow.values());
+
+          insertedRow.update(changeTypeIndex, ChangelogOperation.UPDATE_AFTER.name());
+          this.cachedRow = RowFactory.create(insertedRow.values());
+        } else {
+          // recover the values of change type
+          deletedRow.update(changeTypeIndex, DELETE);
+          insertedRow.update(changeTypeIndex, INSERT);
+          this.cachedRow = nextRow;
+        }
+
+      } else {
+        this.cachedRow = nextRow;
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row currentRow() {
+    if (cachedRow != null) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    } else {
+      return rowIterator.next();
+    }
+  }
+
+  private boolean withinPartition(Row currentRow, Row nextRow) {
+    for (int i = 0; i < partitionIdx.size(); i++) {
```

Comment:
I'm a bit confused here.  What rows are these?  They have directly the partition values in the row?  (No need to transform?).  And the partition field ids are always the same for all rows?

Reply:
Yes. The partition columns here are identifier columns, which will be provided by the CDC procedure in #6012.